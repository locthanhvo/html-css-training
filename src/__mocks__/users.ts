import { STATUS } from '@/constants';

export const USERS = Array.from({ length: 5 }).map((_, index) => ({
  id: `${index}`,
  firstName: 'Devon',
  lastName: 'Lane',
  email: 'devon@mail.com',
  phone: '+237 6 99 88 77 66',
  status: STATUS.VALID,
  createdAt: '2024-06-05 14:46:58',
  updatedAt: '2024-06-05 14:46:58',
}));
