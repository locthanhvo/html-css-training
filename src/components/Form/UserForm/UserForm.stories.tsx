import { StoryObj, Meta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import UserForm from '.';

const meta: Meta<typeof UserForm> = {
  title: 'Components/UserForm',
  tags: ['autodocs'],
  component: UserForm,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof UserForm>;

export const AddUser: Story = {
  args: {},
};

export const EditUser: Story = {
  args: {
    isEdit: true,
  },
};
