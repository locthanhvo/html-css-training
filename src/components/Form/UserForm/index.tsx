import { memo, useCallback } from 'react';
import { Flex } from '@chakra-ui/react';
import { Controller, useForm } from 'react-hook-form';

// Components
import { InputField, ActionButton } from '@/components';

// Types
import { TUser } from '@/types';

// Utils
import { displayFieldUserForm, formatDateString } from '@/utils';
import { STATUS } from '@/constants';

interface UserFormProps {
  isEdit?: boolean;
  isLoading?: boolean;
  initialValues?: TUser;
  onSubmit: (data: TUser) => void;
}

const UserForm = ({
  initialValues,
  isLoading = false,
  isEdit = false,
  onSubmit,
}: UserFormProps) => {
  const {
    control,
    formState: { isSubmitting, isValid, isDirty },
    handleSubmit,
    clearErrors,
  } = useForm<TUser>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    defaultValues: {
      email: initialValues?.email ?? '',
      phone: initialValues?.phone ?? '',
      password: initialValues?.password ?? '',
      firstName: initialValues?.firstName ?? '',
      lastName: initialValues?.lastName ?? '',
      status: initialValues?.status ?? STATUS.VALID,
      ...(isEdit && {
        createdAt:
          formatDateString(initialValues?.createdAt) ||
          formatDateString(new Date().toISOString()),
        updatedAt:
          formatDateString(initialValues?.updatedAt) ||
          formatDateString(new Date().toISOString()),
      }),
    },
  });

  const handleInputChange = useCallback(
    ({
      field,
      data,
      isError,
      onChange,
    }: {
      field: keyof TUser;
      data: string;
      isError: boolean;
      onChange: (value: string) => void;
    }) => {
      isError && clearErrors(field);
      onChange(data);
    },
    [clearErrors],
  );

  return (
    <Flex
      as="form"
      w="full"
      direction="column"
      gap={4}
      onSubmit={handleSubmit(onSubmit)}
    >
      <Flex gap={3}>
        {displayFieldUserForm(isEdit)
          .slice(0, 2)
          .map(({ name, rules, type, label, placeholder, ariaLabel }) => (
            <Controller
              key={name}
              control={control}
              rules={rules}
              name={name}
              render={({ field, fieldState: { error } }) => (
                <InputField
                  type={type}
                  label={label}
                  placeholder={placeholder}
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={(data) =>
                    handleInputChange({
                      field: name,
                      data,
                      isError: !!error,
                      onChange: field.onChange,
                    })
                  }
                  aria-label={ariaLabel}
                />
              )}
            />
          ))}
      </Flex>

      <Flex gap={3}>
        {displayFieldUserForm(isEdit)
          .slice(2, 4)
          .map(({ name, rules, type, label, placeholder, ariaLabel }) => (
            <Controller
              key={name}
              control={control}
              rules={rules}
              name={name}
              render={({ field, fieldState: { error } }) => (
                <InputField
                  type={type}
                  label={label}
                  placeholder={placeholder}
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={(data) =>
                    handleInputChange({
                      field: name,
                      data,
                      isError: !!error,
                      onChange: field.onChange,
                    })
                  }
                  aria-label={ariaLabel}
                />
              )}
            />
          ))}
      </Flex>

      <Flex gap={3}>
        {displayFieldUserForm(isEdit)
          .slice(4, 6)
          .map(({ name, rules, type, label, placeholder, ariaLabel }) => (
            <Controller
              key={name}
              control={control}
              rules={rules}
              name={name}
              render={({ field, fieldState: { error } }) => (
                <InputField
                  type={type}
                  label={label}
                  placeholder={placeholder}
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={true}
                  onChange={(data) =>
                    handleInputChange({
                      field: name,
                      data,
                      isError: !!error,
                      onChange: field.onChange,
                    })
                  }
                  aria-label={ariaLabel}
                />
              )}
            />
          ))}
      </Flex>

      <ActionButton
        type="submit"
        size="lg"
        title={isEdit ? 'Update' : 'Create'}
        bgColor="blue.200"
        isDisabled={!isDirty || (!isValid && !isSubmitting)}
        isLoading={isLoading || isSubmitting}
      />
    </Flex>
  );
};

export default memo(UserForm);
