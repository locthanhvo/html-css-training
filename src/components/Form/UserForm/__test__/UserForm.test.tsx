import { render, screen } from '@testing-library/react';

// Component
import UserForm from '..';
import { BrowserRouter } from 'react-router-dom';
import { USERS } from '@/__mocks__';

const mockOnSubmit = jest.fn();

describe('UserForm components', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('match snapshot with add form', () => {
    const { container } = render(
      <BrowserRouter>
        <UserForm onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('match snapshot with edit form', () => {
    const { container } = render(
      <BrowserRouter>
        <UserForm isEdit initialValues={USERS[0]} onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders Add form by default', () => {
    render(
      <BrowserRouter>
        <UserForm onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(screen.getByRole('button', { name: 'Create' })).toBeInTheDocument();
  });

  it('renders Edit form by default', () => {
    render(
      <BrowserRouter>
        <UserForm isEdit onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(screen.getByRole('button', { name: 'Update' })).toBeInTheDocument();
  });
});
