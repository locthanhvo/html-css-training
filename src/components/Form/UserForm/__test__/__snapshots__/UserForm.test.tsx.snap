// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm components match snapshot with add form 1`] = `
<div>
  <form
    class="css-jbe4xg"
  >
    <div
      class="css-blvggq"
    >
      <div
        aria-label="email"
        class="chakra-form-control css-0"
        name="email"
        placeholder="Email"
        role="group"
        value=""
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Email
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="email"
            class="chakra-input css-adz8kx"
            id="field-:r0:"
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        aria-label="phone"
        class="chakra-form-control css-0"
        name="phone"
        placeholder="Phone"
        role="group"
        value=""
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Phone
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="phone"
            class="chakra-input css-adz8kx"
            id="field-:r1:"
            name="phone"
            placeholder="Phone"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-blvggq"
    >
      <div
        aria-label="firstName"
        class="chakra-form-control css-0"
        name="firstName"
        placeholder="First Name"
        role="group"
        value=""
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          First Name
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="firstName"
            class="chakra-input css-adz8kx"
            id="field-:r2:"
            name="firstName"
            placeholder="First Name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        aria-label="lastName"
        class="chakra-form-control css-0"
        name="lastName"
        placeholder="Last Name"
        role="group"
        value=""
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r3:"
          id="field-:r3:-label"
        >
          Last Name
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="lastName"
            class="chakra-input css-adz8kx"
            id="field-:r3:"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-blvggq"
    />
    <button
      class="chakra-button css-ey593w"
      disabled=""
      name="Create"
      type="submit"
    >
      Create
    </button>
  </form>
</div>
`;

exports[`UserForm components match snapshot with edit form 1`] = `
<div>
  <form
    class="css-jbe4xg"
  >
    <div
      class="css-blvggq"
    >
      <div
        aria-label="email"
        class="chakra-form-control css-0"
        name="email"
        placeholder="Email"
        role="group"
        value="devon@mail.com"
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r4:"
          id="field-:r4:-label"
        >
          Email
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="email"
            class="chakra-input css-adz8kx"
            id="field-:r4:"
            name="email"
            placeholder="Email"
            type="text"
            value="devon@mail.com"
          />
        </div>
      </div>
      <div
        aria-label="phone"
        class="chakra-form-control css-0"
        name="phone"
        placeholder="Phone"
        role="group"
        value="+237 6 99 88 77 66"
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r5:"
          id="field-:r5:-label"
        >
          Phone
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="phone"
            class="chakra-input css-adz8kx"
            id="field-:r5:"
            name="phone"
            placeholder="Phone"
            type="text"
            value="+237 6 99 88 77 66"
          />
        </div>
      </div>
    </div>
    <div
      class="css-blvggq"
    >
      <div
        aria-label="firstName"
        class="chakra-form-control css-0"
        name="firstName"
        placeholder="First Name"
        role="group"
        value="Devon"
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r6:"
          id="field-:r6:-label"
        >
          First Name
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="firstName"
            class="chakra-input css-adz8kx"
            id="field-:r6:"
            name="firstName"
            placeholder="First Name"
            type="text"
            value="Devon"
          />
        </div>
      </div>
      <div
        aria-label="lastName"
        class="chakra-form-control css-0"
        name="lastName"
        placeholder="Last Name"
        role="group"
        value="Lane"
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r7:"
          id="field-:r7:-label"
        >
          Last Name
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="lastName"
            class="chakra-input css-adz8kx"
            id="field-:r7:"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value="Lane"
          />
        </div>
      </div>
    </div>
    <div
      class="css-blvggq"
    >
      <div
        aria-label="createdAt"
        class="chakra-form-control css-0"
        data-disabled=""
        name="createdAt"
        placeholder="Created At"
        role="group"
        value="2024-06-05 14:46:58"
      >
        <label
          class="chakra-form__label css-zecyy3"
          data-disabled=""
          for="field-:r8:"
          id="field-:r8:-label"
        >
          Created At
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="createdAt"
            class="chakra-input css-adz8kx"
            disabled=""
            id="field-:r8:"
            name="createdAt"
            placeholder="Created At"
            type="datetime-local"
            value="2024-06-05 14:46:58"
          />
        </div>
      </div>
      <div
        aria-label="updatedAt"
        class="chakra-form-control css-0"
        data-disabled=""
        name="updatedAt"
        placeholder="Updated At"
        role="group"
        value="2024-06-05 14:46:58"
      >
        <label
          class="chakra-form__label css-zecyy3"
          data-disabled=""
          for="field-:r9:"
          id="field-:r9:-label"
        >
          Updated At
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="updatedAt"
            class="chakra-input css-adz8kx"
            disabled=""
            id="field-:r9:"
            name="updatedAt"
            placeholder="Updated At"
            type="datetime-local"
            value="2024-06-05 14:46:58"
          />
        </div>
      </div>
    </div>
    <button
      class="chakra-button css-ey593w"
      disabled=""
      name="Update"
      type="submit"
    >
      Update
    </button>
  </form>
</div>
`;
